### PR TITLE
Crash Refactor Part I

### PR DESCRIPTION
**Pull request by mnabian**
# PhysicsNeMo Pull Request

## Description

This PR updates the structural mechanics crash example to add support for bumper beam dataset, support for global variables and dynamic features/targets, self-contained experiment configs, etc. It also updates the readme and fixes several bugs. This PR will be followed by a Part 2 PR for refactoring the rollout methods (please disregard the commented out GNN rollout methods in this PR, which will be fixed later).

Validated on the bumper beam dataset (left is prediction, right is ground truth)
image

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

### CODE REVIEW

File: examples/structural_mechanics/crash/vtp_reader.py [hunk 1]
```diff
@@ -174,18 +176,35 @@ def process_vtp_data(data_dir, num_samples=2, write_vtp=False, logger=None):
...
 
+        # Get global features for this run
+        run_id = os.path.splitext(os.path.basename(vtp_path))[0]
+        if global_features_filepath is not None:
+            global_features = get_global_features_for_run(
+                all_global_features,
+                run_id,
+            )
+        else:
+            global_features = {}
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/crash/vtp_reader.py:234]: Fixed

File: examples/structural_mechanics/crash/vtp_reader.py [hunk 2]
```diff
@@ -210,15 +229,19 @@ def process_vtp_data(data_dir, num_samples=2, write_vtp=False, logger=None):
         )
 
         # Create record with coords and all other point data fields
-        record = {"coords": mesh_pos_all}
-        record.update(point_data_dict)  # Add thickness and any other fields
+        record = {
+            "coords": mesh_pos_all,
+            "point_data": point_data_dict,
+        }
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/crash/vtp_reader.py]: Fixed.

File: examples/structural_mechanics/crash/rollout.py [hunk 3]
```diff
@@ -18,14 +18,88 @@
...
+        Returns:
+            [T, N, 3] rollout of predicted positions
+        """
+        inputs = sample.node_features
+        coords = inputs["coords"]  # [N,3]
+        features = inputs.get("features", coords.new_zeros((coords.size(0), 0)))
+        initial_vel = sample.global_features["velocity_x"].item()  # TODO mnabian needs better handling for unification
+        global_features = torch.stack([sample.global_features[k] for k in sample.global_features.keys()], dim=0)
+        N = coords.size(0)
+        device = coords.device
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/crash/rollout.py]: Fixed.

File: examples/structural_mechanics/crash/rollout.py [hunk 4]
```diff
@@ -18,14 +18,88 @@
...
+        Args:
+            sample: SimSample containing node_features and node_target
+            data_stats: dict containing normalization stats
+        Returns:
+            [T, N, 3] rollout of predicted positions
+        """
+        inputs = sample.node_features
+        coords = inputs["coords"]  # [N,3]
+        features = inputs.get("features", coords.new_zeros((coords.size(0), 0)))
+        initial_vel = sample.global_features["velocity_x"].item()  # TODO mnabian needs better handling for unification
```
[the remark this answers is not in the document]
Comment (mnabian) [reply examples/structural_mechanics/crash/rollout.py]: Fixed

File: examples/structural_mechanics/crash/conf/experiment_crash_geotransolver.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    chdir: True
+  run:
+    dir: ./outputs/
+
+experiment_name: "Car-Crash-Transolver"
+experiment_desc: "Car body-in-white crash experiment using Transolver"
+run_desc: "Car body-in-white crash experiment using Transolver"
+
+defaults:
+  - reader: vtp
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/conf/experiment_crash_geotransolver.yaml:34]: `d3plot` for crash


```suggestion
  - reader: d3plot
```

File: examples/structural_mechanics/crash/datapipe.py [hunk 6]
```diff
@@ -125,28 +145,65 @@ def __init__(
...
+                    "datapipe.global_features is configured but training.global_features_filepath "
+                    "is unresolved (???). Set it via config or CLI, e.g. "
+                    "training.global_features_filepath=/path/to/global_features.json"
+                )
+
+        self.srcs, self.dsts, point_data, global_features = reader(
             data_dir=self.data_dir,
             num_samples=num_samples,
             split=split,
+            global_features_filepath=self.global_features_filepath,
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/datapipe.py:188]: d3plot_reader.py don't have `global_features_filepath` as an input argument resulting in error.

File: examples/structural_mechanics/crash/datapipe.py [hunk 7]
```diff
@@ -125,28 +145,65 @@ def __init__(
...
+                    "training.global_features_filepath=/path/to/global_features.json"
+                )
+            if str(global_features_filepath).strip() == "???":
+                raise ValueError(
+                    "datapipe.global_features is configured but training.global_features_filepath "
+                    "is unresolved (???). Set it via config or CLI, e.g. "
+                    "training.global_features_filepath=/path/to/global_features.json"
+                )
+
+        self.srcs, self.dsts, point_data, global_features = reader(
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/datapipe.py:184]: `d3plot_reader.py > process_d3plot_data()` only returns 3 values, and don't return `global_features`

File: examples/structural_mechanics/crash/rollout.py
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/rollout.py:132]: Given that Autoregressive without time as input provide better validation loss, should we remove the `time_t` as input?

File: examples/structural_mechanics/crash/train.py
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/train.py:369]: Should we aggregate loss from all nodes or we can keep it for next PR.

```suggestion

        total_loss_n = torch.tensor([total_loss], dtype=torch.float64, device=dist.device)
        num_batches_n = torch.tensor([num_batches], dtype=torch.float64, device=dist.device)
        if dist.world_size > 1:
            torch.distributed.all_reduce(total_loss_n, op=torch.distributed.ReduceOp.SUM)
            torch.distributed.all_reduce(num_batches_n, op=torch.distributed.ReduceOp.SUM)
        global_avg_loss = (total_loss_n / num_batches_n.clamp(min=1)).item()

        epoch_duration = time.time() - start
        logger0.info(
            f"Epoch {epoch + 1}/{cfg.training.epochs} "
            f"avg_loss: {global_avg_loss:.6f} "
            f"lr: {trainer.optimizer.param_groups[0]['lr']:.3e} "
            f"duration: {epoch_duration:.2f}s"
        )

        if dist.rank == 0:
            trainer.writer.add_scalar("loss", global_avg_loss, epoch)
            trainer.writer.add_scalar(
                "learning_rate", trainer.optimizer.param_groups[0]["lr"], epoch
            )

```

File: examples/structural_mechanics/crash/requirements.txt [hunk 8]
```diff
@@ -1,4 +1,4 @@
-jaxtyping==0.3.3
-lasso-python==2.0.3
-torch_geometric==2.6.1
-torch_scatter>=2.1.2
+pyvista>=0.43.0
+tabulate>=0.9.0
+tensorboard>=2.20.2
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/requirements.txt]: ERROR: No matching distribution found for tensorboard>=2.20.2


```suggestion
tensorboard>=2.20.0
```

File: examples/structural_mechanics/crash/requirements.txt [hunk 9]
```diff
@@ -1,4 +1,4 @@
...
+torchinfo>=1.8
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/requirements.txt:4]: Need lasso for car crash dataset

```suggestion
torchinfo>=1.8
lasso-python==2.0.4
```

File: examples/structural_mechanics/crash/conf/training/default.yaml [hunk 10]
```diff
@@ -16,24 +16,25 @@
...
 # ┌───────────────────────────────────────────┐
 # │                   Data                    │
-# └───────────────────────────────────────────┘  
+# └───────────────────────────────────────────┘
 
-raw_data_dir:  "/code/datasets/gm_crash/train"  # TODO change
-raw_data_dir_validation:  "/code/datasets/gm_crash/validation"  
-max_workers_preprocessing: 64  # Maximum parallel workers
+raw_data_dir: ???               # Required: set via experiment config
+raw_data_dir_validation: ???    # Required: set via experiment config
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/conf/training/default.yaml:22]: What are your thoughts on - 
```suggestion
raw_data_dir: {config.raw_data_dir}               # Required: set via experiment config
raw_data_dir_validation: {config.raw_data_dir_validation}    # Required: set via experiment config
```

File: examples/structural_mechanics/crash/conf/training/default.yaml [hunk 11]
```diff
@@ -16,24 +16,25 @@
...
 # ┌───────────────────────────────────────────┐
 # │                 Training                  │
-# └───────────────────────────────────────────┘  
+# └───────────────────────────────────────────┘
 
-num_time_steps: 14
-num_training_samples: 8
-num_validation_samples: 8
+num_time_steps: ???             # Required: set via experiment config
+num_training_samples: ???       # Required: set via experiment config
```
Comment (dakhare-creator) [inline examples/structural_mechanics/crash/conf/training/default.yaml:31]: What are your thoughts on -

```suggestion
num_time_steps: {config.num_time_steps}            # Required: set via experiment config
num_training_samples: {config.num_training_samples}      # Required: set via experiment config
```

File: examples/structural_mechanics/crash/rollout.py
Comment (mnabian) [reply examples/structural_mechanics/crash/rollout.py:132]: We will revisit the rollout methods in Part 2 of the refactor in a separate PR.

File: examples/structural_mechanics/crash/conf/training/default.yaml (hunk 10 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/conf/training/default.yaml:22]: The suggestion won't work, but the current config already behaves the same way.

File: examples/structural_mechanics/crash/conf/experiment_crash_geotransolver.yaml (hunk 5 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/conf/experiment_crash_geotransolver.yaml:34]: d3plot is fully removed as we now rely on the Curator for data preprocessing.

File: examples/structural_mechanics/crash/datapipe.py (hunk 7 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/datapipe.py:184]: d3plot_reader is not used anymore.

File: examples/structural_mechanics/crash/train.py
Comment (mnabian) [reply examples/structural_mechanics/crash/train.py:369]: Let's keep that for the second PR.

File: examples/structural_mechanics/crash/requirements.txt (hunk 9 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/requirements.txt:4]: d3plot_reader is not used anymore.

File: examples/structural_mechanics/crash/conf/training/default.yaml (hunk 11 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/conf/training/default.yaml:31]: Same here.

File: examples/structural_mechanics/crash/datapipe.py (hunk 6 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/datapipe.py:188]: d3plot_reader is not used anymore.

File: examples/structural_mechanics/crash/datapipe.py [hunk 12]
```diff
@@ -254,16 +359,40 @@ def build_xy(self, idx: int):
         pos_t0 = pos_seq[0]  # [N,3]
         x = {"coords": pos_t0, "features": feats}
 
-        # Flatten all future positions along feature dim
-        y = pos_seq[1:].transpose(0, 1).flatten(start_dim=1)  # [N,(T-1)*3]
+        # Build interleaved per-timestep targets: [dx, dy, dz, eps, vm, ...] per timestep
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/datapipe.py]: No longer needed

File: examples/structural_mechanics/crash/datapipe.py [hunk 13]
```diff
@@ -254,16 +359,40 @@ def build_xy(self, idx: int):
...
+        T_out, Fo = y.shape[1], y.shape[2]
         assert x["coords"].shape == (N, 3) and x["features"].shape == (N, F), (
             f"coords shape {x['coords'].shape}, features shape {x['features'].shape}, expected (N,3)/(N,{F})"
         )
-        assert y.shape == (N, Dout), (
-            f"y shape mismatch: expected {(N, Dout)}, got {y.shape}"
+        assert y.shape == (N, T_out, Fo), (
+            f"target shape {y.shape} does not match expected (N={N}, T={T_out}, Fo={Fo})"
         )
+        # y now includes dynamic targets interleaved per timestep
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/datapipe.py]: No longer needed as well.

File: examples/structural_mechanics/crash/inference.py [hunk 14]
```diff
@@ -263,29 +340,38 @@ def main(cfg: DictConfig):
...
         return
 
-    logger0.info(f"Found {len(run_dirs)} runs under {parent_dir}")
+    logger0.info(f"Found {len(run_items)} runs under {parent_dir}")
 
-    # Shard run list across ranks: rank r processes run_dirs[r::world_size]
-    my_runs = run_dirs[dist.rank :: dist.world_size]
-    logger.info(f"[Rank {dist.rank}] Assigned {len(my_runs)} runs.")
+    # Shard run list across ranks: rank r processes run_items[r::world_size]
+    my_items = run_items[dist.rank :: dist.world_size]
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/inference.py:364]: Just verify if this works if run_items is not perfectly divisible by world_size

File: examples/structural_mechanics/crash/README.md [hunk 15]
```diff
@@ -11,104 +11,74 @@ Machine Learning (ML) surrogates provide a promising alternative by learning map
...
+   **Training:**
 
-5) Model config: ensure input dimensions match your features.
+   ```bash
+   # Single GPU
+   python train.py --config-name=experiment_bumper_geotransolver
 
-- Transolver (time‑conditional): set `functional_dim = len(features)` and `embedding_dim = 3`;
+   # Multi-GPU (DDP)
+   torchrun --standalone --nproc_per_node=4 train.py --config-name=experiment_bumper_geotransolver
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/README.md]: Maybe we can remove --standalone so that this same command also works for multi-node env?

File: examples/structural_mechanics/crash/README.md [hunk 16]
```diff
@@ -11,104 +11,74 @@ Machine Learning (ML) surrogates provide a promising alternative by learning map
...
-functional_dim: 1    # e.g., 1 if features: [thickness]
-embedding_dim: 3
-time_input: true
-```
+   **Inference:**
 
-6) Launch training.
+   ```bash
+   python inference.py --config-name=experiment_bumper_geotransolver
+   ```
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/README.md:73]: Maybe document the multi-GPU inference too?

File: examples/structural_mechanics/crash/README.md [hunk 17]
```diff
@@ -221,79 +188,165 @@ This format is directly compatible with the Zarr reader in this example.
...
 ```bash
-python train.py
+python train.py --config-name=experiment_bumper_geotransolver
 ```
 
 Multi-GPU (Distributed Data Parallel):
 
 ```bash
-torchrun --standalone --nproc_per_node=<NUM_GPUS> train.py
+torchrun --standalone --nproc_per_node=<NUM_GPUS> train.py --config-name=experiment_bumper_geotransolver
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/README.md]: Same comment here

File: examples/structural_mechanics/crash/README.md [hunk 18]
```diff
@@ -327,13 +380,80 @@ For completeness, the datapipe also records a lightweight name-to-column map cal
...
+```yaml
+# conf/model/geotransolver_one_shot_rollout_training.yaml
+global_dim: 3   # must match len(datapipe.global_features)
+```
+
+If `global_features` is `null`, `sample.global_features` is `None` and the model must handle this case (currently only `GeoTransolverOneShotTraining` uses global features; other models ignore them).
+
+## Reader: built-in VTP and Zarr readers and how to add your own
+
+The reader opens preprocessed simulation data and produces the arrays the datapipe consumes. Raw LS-DYNA d3plot files must be preprocessed to VTP or Zarr using [PhysicsNeMo-Curator](https://github.com/NVIDIA/physicsnemo-curator) before use. The reader is swappable via Hydra so you can adapt the pipeline to different formats or add your own.
```
Comment (ktangsali) [inline examples/structural_mechanics/crash/README.md]: Maybe point to the specific link rather than the repo.

File: examples/structural_mechanics/crash/datapipe.py (hunk 12 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/datapipe.py]: Removed.

File: examples/structural_mechanics/crash/datapipe.py (hunk 13 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/datapipe.py]: Removed.

File: examples/structural_mechanics/crash/inference.py (hunk 14 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/inference.py:364]: This gives round-robin partitioning, and it behaves well even if un_items is not perfectly divisible by world_size.

File: examples/structural_mechanics/crash/README.md (hunk 15 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/README.md]: Removed.

File: examples/structural_mechanics/crash/README.md (hunk 16 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/README.md:73]: Added.

File: examples/structural_mechanics/crash/README.md (hunk 17 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/README.md]: Removed.

File: examples/structural_mechanics/crash/README.md (hunk 18 above)
Comment (mnabian) [reply examples/structural_mechanics/crash/README.md]: Great comment! Fixed it.
